akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -437,6 +463,56 @@ private[projection] class R2dbcOffsetStore(
       }
     }
   }
+
+  def isTooFarAhead(offsets: immutable.IndexedSeq[OffsetPidSeqNr]): Future[Boolean] = {
+    if (offsets.isEmpty || !offsets.head.offset.isInstanceOf[TimestampOffset])
+      FutureFalse
+    else {
+      val latestTimestampInOffsets = offsets.iterator.map {
+        case OffsetPidSeqNr(t: TimestampOffset, _) => t.timestamp
+        case _ =>
+          throw new IllegalArgumentException(
+            "Mix of TimestampOffset and other offset type in same transaction is not supported")
+      }.max
+
+      val oldState = state.get()
+
+      // FIXME config, related to backtracking window
+      // temporary value of 10s to match ChangeSliceRangesSpec
+      val acceptedDifference = JDuration.ofSeconds(10)
+
+      val stateWithUpdatedLatestTimestampBySlice =
+        if (JDuration
+              .between(oldState.latestTimestampBySlice.readTimestamp, latestTimestampInOffsets)
+              .compareTo(JDuration.ofSeconds(1)) > 0) { // FIXME config, related to acceptedDifference
+          dao.readLatestTimestampFromOtherSlices().map { timestampsFromOtherSlices =>
+            val timestampsFromMySlices = oldState.latestBySlice.iterator.map(r => r.slice -> r.timestamp).toMap
+            val newLatestTimestampBySlice =
+              LatestTimestampBySlice(latestTimestampInOffsets, timestampsFromOtherSlices ++ timestampsFromMySlices)
+            logger.debug("{}", newLatestTimestampBySlice) // FIXME remove
+            oldState.copy(latestTimestampBySlice = newLatestTimestampBySlice)
+          }
+        } else {
+          Future.successful(oldState)
+        }
+
+      stateWithUpdatedLatestTimestampBySlice.map { newState =>
+        if (!state.compareAndSet(oldState, newState))
+          throw new IllegalStateException("Unexpected concurrent modification of state from isTooFarAhead.")
+        newState.latestTimestampBySlice.earliest match {
+          case None           => false
+          case Some(earliest) =>
+            // FIXME config instead of half timeWindow?
+            val betweenEarliestAndLatest = JDuration.between(earliest, latestTimestampInOffsets)
+            logger.debug("Difference to earliest [{} ms]", betweenEarliestAndLatest.toMillis) // FIXME remove
+//            if (betweenEarliestAndLatest.compareTo(acceptedDifference) > 0)
+//              throw new RuntimeException("Too far ahead") // FIXME real exception
+            betweenEarliestAndLatest.compareTo(acceptedDifference.dividedBy(2)) > 0
```

Comment:
This doesn't really work out because there could easily be a huge jump between a new offset and some earlier offset, because there might simply not have been any events in between. That's maybe the easiest way to trigger the downscaling problem. I wrote the test like that.

Reply:
We also have the `timestamp_consumed` (TimestampOffset.readTimestamp), but it would be the same problem for a new event arriving after an idle period.